Expose fragment ranges for testing.

megangardner opened the pull request:
#### f13a765177da98919f5857208c0606f5193f7d55
<pre>
Expose fragment ranges for testing.
<a href="https://bugs.webkit.org/show_bug.cgi?id=292871">https://bugs.webkit.org/show_bug.cgi?id=292871</a>
<a href="https://rdar.apple.com/151151068">rdar://151151068</a>

Reviewed by Wenson Hsieh.

We would like to build some more extensive testing
tools that ensure that fragment generation is working
in all cases. We need to verify that the fragments
are in the correct location, so adding this SPI to
expose their locations in the document.

* Source/WebKit/UIProcess/API/Cocoa/WKWebView.mm:
(-[WKWebView _textFragmentRangesWithCompletionHandler:]):
* Source/WebKit/UIProcess/API/Cocoa/WKWebViewPrivate.h:
* Source/WebKit/UIProcess/Cocoa/WebPageProxyCocoa.mm:
(WebKit::WebPageProxy::createTextFragmentDirectiveFromSelection):
(WebKit::WebPageProxy::getTextFragmentRanges):
* Source/WebKit/UIProcess/WebPageProxy.h:
* Source/WebKit/WebProcess/WebPage/WebPage.cpp:
(WebKit::WebPage::getTextFragmentRanges):
* Source/WebKit/WebProcess/WebPage/WebPage.h:
* Source/WebKit/WebProcess/WebPage/WebPage.messages.in:
* Tools/TestWebKitAPI/Tests/WebKitCocoa/WKFragmentDirectiveGeneration.mm:
(TestWebKitAPI::TEST(FragmentDirectiveGeneration, VerifyFragmentRanges)):

Canonical link: <a href="https://commits.webkit.org/294911@main">https://commits.webkit.org/294911@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/40952caf8b8b3dc5a59e2525b5427f4846034f3b

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/103497 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/23179 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/13499 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/108673 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/54142 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/105536 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/23530 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/31730 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/78652 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [⏳ 🧪 win-tests](https://ews-build.webkit.org/#/builders/Win-Tests-EWS "Waiting to run tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/106503 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/18248 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/93363 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/58986 "Passed tests") | 
| | [  ~~🧪 ios-wk2-wpt~~](https://ews-build.webkit.org/#/builders/133/builds/18068 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/11410 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/53498 "Built successfully") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/13/builds/87859 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/11470 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/111051 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/30644 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/122/builds/22605 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/87648 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/31005 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/89563 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/87289 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/22217 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/32166 "Passed tests") | [  ~~🧪 mac-intel-wk2~~](https://ews-build.webkit.org/#/builders/137/builds/9876 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/24995 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/30572 "Built successfully") | [  ~~🛠 mac-safer-cpp~~](https://ews-build.webkit.org/#/builders/120/builds/35884 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/30372 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/33703 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/31933 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

whsieh commented on Source/WebKit/UIProcess/API/Cocoa/WKWebViewPrivate.h:
```diff
@@ -152,6 +152,8 @@ typedef NS_ENUM(NSInteger, _WKImmediateActionType) {
 @class NSImage;
 #endif
 
+typedef struct _NSRange NSRange;
```
Is this forward declaration necessary? I think we should already have pulled in all public Foundation headers here, since we have AppKit/UIKit.

whsieh commented on Source/WebKit/UIProcess/API/Cocoa/WKWebViewPrivate.h:
```diff
@@ -454,6 +456,8 @@ for this property.
 
 - (void)_textFragmentDirectiveFromSelectionWithCompletionHandler:(void(^)(NSURL *))completionHandler WK_API_AVAILABLE(macos(WK_MAC_TBA), ios(WK_IOS_TBA), visionos(WK_XROS_TBA));
 
+- (void)_textFragmentRangesWithCompletionHandler:(void(^)(NSArray<NSValue *> * fragmentRanges))completionHandler WK_API_AVAILABLE(macos(WK_MAC_TBA), ios(WK_IOS_TBA), visionos(WK_XROS_TBA));
```
If it's really only for testing, I'd recommend putting it in `WKWebViewPrivateForTesting.h` (optionally, with a `-ForTesting` suffix in the name).

whsieh commented on Source/WebKit/WebProcess/WebPage/WebPage.cpp:
```diff
@@ -9414,6 +9414,28 @@ void WebPage::createTextFragmentDirectiveFromSelection(CompletionHandler<void(UR
+        return;
+
+    Vector<EditingRange> editingRanges;
+    if (auto highlightRegistry = document->fragmentHighlightRegistryIfExists()) {
```
```suggestion
    if (RefPtr highlightRegistry = document->fragmentHighlightRegistryIfExists()) {
```

whsieh commented on Source/WebKit/WebProcess/WebPage/WebPage.cpp:
```diff
@@ -9414,6 +9414,28 @@ void WebPage::createTextFragmentDirectiveFromSelection(CompletionHandler<void(UR
+{
+    RefPtr focusedOrMainFrame = protectedCorePage()->checkedFocusController()->focusedOrMainFrame();
+    if (!focusedOrMainFrame)
+        return;
```
You should still call the completion handler before early returning:

```suggestion
    if (!focusedOrMainFrame) {
        completionHandler({ });
        return;
    }
```

whsieh commented on Source/WebKit/WebProcess/WebPage/WebPage.cpp:
```diff
@@ -9414,6 +9414,28 @@ void WebPage::createTextFragmentDirectiveFromSelection(CompletionHandler<void(UR
+
+    RefPtr frame = document->frame();
+    if (!frame)
+        return;
```
Ditto:

```suggestion
    RefPtr frame = document->frame();
    if (!frame) {
        completionHandler({ });
        return;
    }
```

whsieh commented on Source/WebKit/WebProcess/WebPage/WebPage.cpp:
```diff
@@ -9414,6 +9414,32 @@ void WebPage::createTextFragmentDirectiveFromSelection(CompletionHandler<void(UR
+    if (RefPtr highlightRegistry = document->fragmentHighlightRegistryIfExists()) {
+        for (auto& highlight : highlightRegistry->map()) {
+            for (auto& highlightRange : highlight.value->highlightRanges())
+                editingRanges.append(EditingRange::fromRange(*frame, makeSimpleRange(highlightRange->range())));
```
Something like:

```suggestion
                editingRanges.append(EditingRange::fromRange(*frame, makeSimpleRange(Ref { highlightRange->range() }.get())));
```

...or you could add a `protectedRange()` getter to `HighlightRange` that returns a `Ref<AbstractRange>`

whsieh commented on Source/WebKit/UIProcess/API/Cocoa/WKWebViewTesting.mm:
```diff
@@ -1004,6 +1004,19 @@ - (void)_setMediaVolumeForTesting:(float)mediaVolume
+            if (resultRange.location != NSNotFound)
+                [resultRanges addObject:[NSValue valueWithRange:resultRange]];
+        }
+        completion(WTFMove(resultRanges.get()));
```
No `WTFMove` here, since `resultRanges` isn't being passed in as an rvalue. `editingRanges` should also be either `const Vector&` or `Vector&&`:

```suggestion
    _page->getTextFragmentRanges([completion = makeBlockPtr(completionHandler)](const Vector<WebKit::EditingRange>& editingRanges) {
        RetainPtr<NSMutableArray<NSValue *>> resultRanges = [NSMutableArray array];
        for (auto editingRange : editingRanges) {
            NSRange resultRange = editingRange;
            if (resultRange.location != NSNotFound)
                [resultRanges addObject:[NSValue valueWithRange:resultRange]];
        }
        completion(resultRanges.get());
```

whsieh reviewed (commented):
(This should fix the Safer C++ failures)

pxlcoder commented on Source/WebKit/UIProcess/API/Cocoa/WKWebViewPrivateForTesting.h:
```diff
@@ -156,6 +156,8 @@ struct WKAppPrivacyReportTestingData {
 
 - (void)_setMediaVolumeForTesting:(float)volume;
 
+- (void)_textFragmentRangesWithCompletionHandlerForTesting:(void(^)(NSArray<NSValue *> * fragmentRanges))completionHandler WK_API_AVAILABLE(macos(WK_MAC_TBA), ios(WK_IOS_TBA), visionos(WK_XROS_TBA));
```
```suggestion
- (void)_textFragmentRangesWithCompletionHandlerForTesting:(void(^)(NSArray<NSValue *> *fragmentRanges))completionHandler WK_API_AVAILABLE(macos(WK_MAC_TBA), ios(WK_IOS_TBA), visionos(WK_XROS_TBA));
```

pxlcoder commented on Source/WebKit/UIProcess/API/Cocoa/WKWebViewTesting.mm:
```diff
@@ -1004,6 +1004,19 @@ - (void)_setMediaVolumeForTesting:(float)mediaVolume
     };
 }
 
+- (void)_textFragmentRangesWithCompletionHandlerForTesting:(void(^)(NSArray<NSValue *> * fragmentRanges))completionHandler
```
```suggestion
- (void)_textFragmentRangesWithCompletionHandlerForTesting:(void(^)(NSArray<NSValue *> *fragmentRanges))completionHandler
```

megangardner committed:
Expose fragment ranges for testing.
https://bugs.webkit.org/show_bug.cgi?id=292871
rdar://151151068

Reviewed by Wenson Hsieh.

We would like to build some more extensive testing
tools that ensure that fragment generation is working
in all cases. We need to verify that the fragments
are in the correct location, so adding this SPI to
expose their locations in the document.

* Source/WebKit/UIProcess/API/Cocoa/WKWebView.mm:
(-[WKWebView _textFragmentRangesWithCompletionHandler:]):
* Source/WebKit/UIProcess/API/Cocoa/WKWebViewPrivate.h:
* Source/WebKit/UIProcess/Cocoa/WebPageProxyCocoa.mm:
(WebKit::WebPageProxy::createTextFragmentDirectiveFromSelection):
(WebKit::WebPageProxy::getTextFragmentRanges):
* Source/WebKit/UIProcess/WebPageProxy.h:
* Source/WebKit/WebProcess/WebPage/WebPage.cpp:
(WebKit::WebPage::getTextFragmentRanges):
* Source/WebKit/WebProcess/WebPage/WebPage.h:
* Source/WebKit/WebProcess/WebPage/WebPage.messages.in:
* Tools/TestWebKitAPI/Tests/WebKitCocoa/WKFragmentDirectiveGeneration.mm:
(TestWebKitAPI::TEST(FragmentDirectiveGeneration, VerifyFragmentRanges)):

Canonical link: https://commits.webkit.org/294911@main
